### PR TITLE
[20307] Ignore custom PIDs defined in Fast DDS when they are received from other vendors

### PR DESCRIPTION
**Pull request by EduPonz**
## Description

With this PR, Fast DDS ignores all custom PIDs defined in Fast DDS when they are received from other vendors.

## Contributor Checklist
- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [x] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- [x] Any new/modified methods have been properly documented using Doxygen. 
- [ ] Changes are ABI compatible: **This is OK as this PR is in preparation for v2.14.0** 
- [x] Changes are API compatible. 
- _N/A_: New feature has been added to the `versions.md` file (if applicable).
- _N/A_: New feature has been documented/Current behavior is correctly described in the documentation. 
    
- _N/A_: Applicable backports have been included in the description.

## Reviewer Checklist
- [x] The PR has a milestone assigned.
- [x] The title and description correctly express the PR's purpose.
- [x] Check contributor checklist is correct.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

**Comment by EduPonz** [comment]
@richiprosima please test this

**Comment by EduPonz** [comment]
@richiprosima please test mac

File: test/unittest/rtps/common/VendorIdTests.cpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
+// Copyright 2022 Proyectos y Sistemas de Mantenimiento SL (eProsima).
```
Comment (MiguelCompany) [inline test/unittest/rtps/common/VendorIdTests.cpp:1]:
```suggestion
// Copyright 2024 Proyectos y Sistemas de Mantenimiento SL (eProsima).
```

File: src/cpp/rtps/builtin/data/ParticipantProxyData.cpp [hunk 2]
```diff
@@ -388,9 +391,10 @@ bool ParticipantProxyData::readFromCDRMessage(
         bool use_encapsulation,
         const NetworkFactory& network,
         bool is_shm_transport_available,
-        bool should_filter_locators)
+        bool should_filter_locators,
+        fastdds::rtps::VendorId_t source_vendor_id)
 {
-    auto param_process = [this, &network, &is_shm_transport_available, &should_filter_locators](
+    auto param_process = [this, &network, &is_shm_transport_available, &should_filter_locators, &source_vendor_id](
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/data/ParticipantProxyData.cpp:397]:
```suggestion
    auto param_process = [this, &network, &is_shm_transport_available, &should_filter_locators, source_vendor_id](
```

File: src/cpp/rtps/builtin/data/ReaderProxyData.cpp [hunk 3]
```diff
@@ -841,6 +843,15 @@ bool ReaderProxyData::readFromCDRMessage(
                     }
                     case fastdds::dds::PID_NETWORK_CONFIGURATION_SET:
                     {
+                        // Ignore custom PID when coming from other vendors except RTI Connext
+                        if ((c_VendorId_eProsima != source_vendor_id) &&
+                                (fastdds::rtps::c_VendorId_rti_connext != source_vendor_id))
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/data/ReaderProxyData.cpp:848]:
```suggestion
                        // Ignore custom PID when coming from other vendors
                        if (c_VendorId_eProsima != source_vendor_id)
```

File: src/cpp/rtps/builtin/data/WriterProxyData.cpp [hunk 4]
```diff
@@ -845,6 +857,15 @@ bool WriterProxyData::readFromCDRMessage(
                     }
                     case fastdds::dds::PID_NETWORK_CONFIGURATION_SET:
                     {
+                        // Ignore custom PID when coming from other vendors except RTI Connext
+                        if ((c_VendorId_eProsima != source_vendor_id) &&
+                                (fastdds::rtps::c_VendorId_rti_connext != source_vendor_id))
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/data/WriterProxyData.cpp:862]:
```suggestion
                        // Ignore custom PID when coming from other vendors
                        if (c_VendorId_eProsima != source_vendor_id)
```

File: src/cpp/fastdds/core/policy/ParameterList.cpp [hunk 5]
```diff
@@ -64,6 +66,18 @@ bool ParameterList::updateCacheChangeFromInlineQos(
                     case PID_CUSTOM_RELATED_SAMPLE_IDENTITY:
                     case PID_RELATED_SAMPLE_IDENTITY:
                     {
+                        // TODO(eduponz): This check is done here because an implicit fall through rises a warning.
+                        // C++17 included a [[fallthrough]] attribute to avoid this kind of warning.
+                        if (pid == PID_RELATED_ENTITY_GUID)
```
Comment (MiguelCompany) [inline src/cpp/fastdds/core/policy/ParameterList.cpp:71]:
```suggestion
                        if (pid == PID_CUSTOM_RELATED_ENTITY_GUID)
```

File: src/cpp/rtps/builtin/data/ReaderProxyData.cpp [hunk 6]
```diff
@@ -978,6 +989,14 @@ bool ReaderProxyData::readFromCDRMessage(
 
                     case fastdds::dds::PID_DISABLE_POSITIVE_ACKS:
                     {
+                        // Ignore custom PID when coming from other vendors
+                        if (c_VendorId_eProsima != source_vendor_id)
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/data/ReaderProxyData.cpp:993]: We also interoperate with RTI connext in this one

File: src/cpp/rtps/builtin/data/WriterProxyData.cpp [hunk 7]
```diff
@@ -950,6 +971,14 @@ bool WriterProxyData::readFromCDRMessage(
                     }
                     case fastdds::dds::PID_DISABLE_POSITIVE_ACKS:
                     {
+                        // Ignore custom PID when coming from other vendors
+                        if (c_VendorId_eProsima != source_vendor_id)
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/data/WriterProxyData.cpp:975]: We also interoperate with RTI connext in this one

File: src/cpp/rtps/messages/MessageReceiver.cpp [hunk 8]
```diff
@@ -1145,15 +1145,16 @@ bool MessageReceiver::proc_Submsg_Heartbeat(
...
+            [was_decoded, &writerGUID, &HBCount, &firstSN, &lastSN, finalFlag, livelinessFlag, this](RTPSReader* reader)
             {
                 // Only used when HAVE_SECURITY is defined
                 static_cast<void>(was_decoded);
 #if HAVE_SECURITY
                 if (was_decoded || !reader->getAttributes().security_attributes().is_submessage_protected)
 #endif  // HAVE_SECURITY
                 {
-                    reader->processHeartbeatMsg(writerGUID, HBCount, firstSN, lastSN, finalFlag, livelinessFlag);
+                    reader->processHeartbeatMsg(writerGUID, HBCount, firstSN, lastSN, finalFlag, livelinessFlag,
```
Comment (MiguelCompany) [inline src/cpp/rtps/messages/MessageReceiver.cpp:1156]: Please look for other calls to this method. I',m pretty sure we are calling this directly for intra-process communications.

Same applies to the other methods where the vendor id argument has been added

File: test/unittest/rtps/builtin/BuiltinDataSerializationTests.cpp [hunk 9]
```diff
@@ -427,42 +436,238 @@ TEST(BuiltinDataSerializationTests, property_list_with_binary_properties)
     EXPECT_NO_THROW(EXPECT_TRUE(out.readFromCDRMessage(&msg, true, network, false, true)));
 }
 
-// Regression test for redmine ticket 20306
+// Regression test for redmine tickets 20306 and 20307
 TEST(BuiltinDataSerializationTests, other_vendor_parameter_list_with_custom_pids)
```
Comment (MiguelCompany) [inline test/unittest/rtps/builtin/BuiltinDataSerializationTests.cpp:440]: Would be nice to have a test for the PIDs interoperable with RTI connext

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: src/cpp/rtps/DataSharing/DataSharingListener.cpp [hunk 10]
```diff
@@ -173,7 +173,8 @@ void DataSharingListener::process_new_data ()
                 {
                     EPROSIMA_LOG_WARNING(RTPS_READER, "GAP (" << last_sequence + 1 << " - " << ch.sequenceNumber - 1 << ")"
                                                               << " detected on datasharing writer " << pool->writer());
-                    reader_->processGapMsg(pool->writer(), last_sequence + 1, SequenceNumberSet_t(ch.sequenceNumber));
+                    reader_->processGapMsg(pool->writer(), last_sequence + 1, SequenceNumberSet_t(
+                                ch.sequenceNumber), c_VendorId_eProsima);
```
Comment (MiguelCompany) [inline src/cpp/rtps/DataSharing/DataSharingListener.cpp:177]:
```suggestion
                    reader_->processGapMsg(pool->writer(), last_sequence + 1,
                                SequenceNumberSet_t(ch.sequenceNumber), c_VendorId_eProsima);
```

File: src/cpp/rtps/DataSharing/DataSharingListener.cpp [hunk 11]
```diff
@@ -182,7 +183,7 @@ void DataSharingListener::process_new_data ()
                                                                            << " detected on datasharing writer " <<
                             pool->writer());
                     reader_->processGapMsg(pool->writer(), SequenceNumber_t(0, 1), SequenceNumberSet_t(
-                                ch.sequenceNumber));
+                                ch.sequenceNumber), c_VendorId_eProsima);
```
Comment (MiguelCompany) [inline src/cpp/rtps/DataSharing/DataSharingListener.cpp:186]: Do the same as in my previous suggestion, please

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

**Review by MiguelCompany — APPROVED** [review]

**Comment by EduPonz** [comment]
@Mergifyio rebase

**Comment by EduPonz** [comment]
@Mergifyio rebase master

**Comment by EduPonz** [comment]
@richiprosima please test this

**Review by MiguelCompany — APPROVED** [review]

**Pull request merged**